In [1]:
# !pip install numpy==1.21.4
# !pip install pandas==1.3.4
# !pip install tensorflow
# !pip install keras==2.1.6
# !pip install scikit-learn

In [2]:
import pandas as pd
import numpy as np

In [3]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [5]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

In [6]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [7]:
n_cols = predictors_norm.shape[1] # number of predictors

In [8]:
import keras

In [9]:
from keras.models import Sequential
from keras.layers import Dense

In [10]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [11]:
# build the model
model = regression_model()

In [12]:
import sklearn.metrics as metrics
from sklearn.model_selection import train_test_split

In [13]:
mse = []

# fit the model
for i in range(50):
    x_train, x_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(x_train, y_train, epochs=50, verbose=2)
    predictions = model.predict(x_test)
    err = metrics.mean_squared_error(predictions, y_test)
    mse.append(err)
    


Epoch 1/50
23/23 - 1s - loss: 12258.7432 - 1s/epoch - 58ms/step
Epoch 2/50
23/23 - 0s - loss: 2464.0508 - 51ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 802.2492 - 48ms/epoch - 2ms/step
Epoch 4/50
23/23 - 0s - loss: 447.9088 - 53ms/epoch - 2ms/step
Epoch 5/50
23/23 - 0s - loss: 328.1961 - 55ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 281.7426 - 47ms/epoch - 2ms/step
Epoch 7/50
23/23 - 0s - loss: 260.4298 - 48ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 247.9681 - 46ms/epoch - 2ms/step
Epoch 9/50
23/23 - 0s - loss: 241.1143 - 49ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 233.9608 - 46ms/epoch - 2ms/step
Epoch 11/50
23/23 - 0s - loss: 229.1321 - 50ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 223.8538 - 38ms/epoch - 2ms/step
Epoch 13/50
23/23 - 0s - loss: 218.8685 - 51ms/epoch - 2ms/step
Epoch 14/50
23/23 - 0s - loss: 215.2939 - 50ms/epoch - 2ms/step
Epoch 15/50
23/23 - 0s - loss: 209.7943 - 49ms/epoch - 2ms/step
Epoch 16/50
23/23 - 0s - loss: 205.8637 - 47ms/

In [14]:
print('Mean of 50 MSEs: ', np.mean(mse))
print('Standard Deviation of 50 MSEs: ', np.std(mse))

Mean of 50 MSEs:  47.16805212258582
Standard Deviation of 50 MSEs:  13.503846122249811


Mean for Part B = 54.12572559325684

Mean for Part D = 47.16805212258582

There is a decrease in mean of MSE's in Part C by 6.957673470671018